pulp_rpm/tests/functional/api/test_sync.py

```diff
@@ -75,1390 +66,988 @@
     SRPM_UNSIGNED_FIXTURE_PACKAGE_COUNT,
     SRPM_UNSIGNED_FIXTURE_URL,
 )
-from pulp_rpm.tests.functional.utils import (
-    gen_rpm_client,
-    gen_rpm_remote,
-    skip_if,
-)
+from pulp_rpm.tests.functional.utils import gen_rpm_remote
 from pulp_rpm.tests.functional.utils import set_up_module as setUpModule  # noqa:F401
 
-from pulpcore.client.pulp_rpm import (
-    ContentDistributionTreesApi,
-    ContentPackagesApi,
-    RepositoriesRpmApi,
-    RpmRepositorySyncURL,
-    RemotesRpmApi,
-    PublicationsRpmApi,
-)
+from pulpcore.client.pulp_rpm import RpmRepositorySyncURL
 from pulpcore.client.pulp_rpm.exceptions import ApiException
 
 
-class BasicSyncTestCase(PulpTestCase):
-    """Sync a repository with the rpm plugin."""
+@pytest.fixture
+def init_and_sync(gen_object_with_cleanup, monitor_task, rpm_repository_api, rpm_rpmremote_api):
+    def _init_and_sync(
+        repository=None,
+        remote=None,
+        url=RPM_UNSIGNED_FIXTURE_URL,
+        policy="immediate",
+        sync_policy="additive",
+        skip_types=None,
+    ):
+        """Initialize a new repository and remote and sync the content from the passed URL."""
+        if repository is None:
+            repository = gen_object_with_cleanup(rpm_repository_api, gen_repo())
+        if remote is None:
+            remote = gen_object_with_cleanup(
+                rpm_rpmremote_api, gen_rpm_remote(url=url, policy=policy)
+            )
 
-    @classmethod
-    def setUpClass(cls):
-        """Create class-wide variables."""
-        cls.client = gen_rpm_client()
-        cls.repo_api = RepositoriesRpmApi(cls.client)
-        cls.remote_api = RemotesRpmApi(cls.client)
-        cls.packages_api = ContentPackagesApi(cls.client)
+        repository_sync_data = RpmRepositorySyncURL(
+            remote=remote.pulp_href, sync_policy=sync_policy, skip_types=skip_types
+        )
+        sync_response = rpm_repository_api.sync(repository.pulp_href, repository_sync_data)
+        monitor_task(sync_response.task)
 
-        delete_orphans()
+        repository = rpm_repository_api.read(repository.pulp_href)
+        return repository, remote
 
-    def test_sync(self):
-        """Sync repositories with the rpm plugin.
+    return _init_and_sync
```

Comment:
How does the fixture work, is it executed fresh every time or only once per set of input arguments?  Basically how is passing in a fixture and calling it different from just calling a function.

Reply:
The fixture initialization is executed once per test case (i.e., when a test case that uses the fixture is run). Then, the "inner" function can be called multiple times within the same test case. I am thinking about changing the frequency of initialization to 1 -> by decorating it with, e.g., `@pytest.fixture(scope="module")`.

There are a couple of advantages over using an ordinary function. One advantage is that this fixture can be later moved to the global conftest.py file if we will find common workflows in other test modules and we will not have to deal with imports explicitly. Another benefit is that teardown methods for generating repositories and remotes are called automatically after finalizing a test case.

Reply:
Now, there is @pytest.fixture(scope="class"). This is the only available option I have because of "gen_object_with_cleanup".